Fix 'task name is not templated in retry callback' (add task_name property to TaskResult)

alikins opened the pull request:
Fixes #18236

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->

 - Bugfix Pull Request


##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
lib/ansible/executor/task_executor.py
lib/ansible/plugins/callback/default.py

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.3.0 (retry_fail_18236 b968b2964e) last updated 2017/02/09 15:46:04 (GMT -400)
  config file = /home/adrian/.ansible.cfg
  configured module search path = [u'/home/adrian/ansible/my-modules', u'/home/adrian/ansible/lib/modules/']

```

alikins committed:
Fix 'task name is not templated in retry callback'

Add a task_name property to TaskResult that knows to
check in TaskResult._task_fields.

Add integration test for v2_retry_runner callback

Fixes #18236

bcoca commented on lib/ansible/executor/task_executor.py:
```diff
@@ -564,6 +564,9 @@ def _evaluate_failed_when_result(result):
                         result['_ansible_retry'] = True
                         result['retries'] = retries
+                        # Add the original task name into the result, so TaskResult can consult it for its name
+                        result['task_name'] = wrap_var(self._task.get_name())
```
_ansible_task_name? to avoid collisions

bcoca replied:
doesn't the result object already get a _task?

alikins replied:
It does, but for v2_runner_retry, it's a task before post_validate happens, so the result._task doesnt have the task name templated.

alikins commented:
@jimi-c updated to include a serializable copy of the Task into '_ansible_result_task' item in the results dict.

bcoca commented:
do we really want the whole object just to get the name?

alikins commented:
@bcoca changed to include all of it based on a comment from jimi-c. But looking close it could potentially just include fields that could/did change in post_validate

alikins commented:
Not sure how to detect when fields would be useful. For a Task(), nothing is always_post_validate. But Base._post_validate excludes Task

alikins commented:
Updated to take advantages of the changes in 3d65482927e33daa30bb6bfc2153b4c64775ee15

sm4rk0 commented on test/integration/Makefile:
```diff
@@ -326,3 +326,6 @@ test_lookup_properties:
 args:
 	(cd targets/args && ./runme.sh $(TEST_FLAGS))
+
+test_callback_retry_task_name:
```
This won't be run automatically as it's not included in `all`or `other` target. Is this intentional?

alikins replied:
Added a test/integration/targets/*/aliases file, so ansible-test will find the tests and run it from CI and locally.

alikins committed:
Add test_callback_retry_task_name to other intg target

alikins committed:
Add ci aliases file.

Join posix/ci/group3. It is the best group. Your favorite
group is not as good as group3. group3 can beat group2 at
thumb wrestling.
